docs/job-writing-guide.md

```diff
@@ -0,0 +1,750 @@
+called `map`, which takes a single callback argument.
+
+Array.map will iterate over every item in the array, invoke your callback
+function with it, save the result toa new array, and when it's finished, it will
+return that array.
+
+```js
+const array = ['a', 'b', 'c'];
+const result = array.map(item => {
+  return item.toUpperCase();
+});
+console.log(array); // ['a', 'b', 'c'];
+console.log(result); // ['A', 'B', 'C'];
+```
+
+Because functions are data in Javascript, we can we-write that code like this
+(which might be a bit more readable)
+
+```js
+const array = ['a', 'b', 'c'];
+const upperCase = item => {
+  return item.toUpperCase();
+};
+const result = array.map(upperCase);
+console.log(array); // ['a', 'b', 'c'];
+console.log(result); // ['A', 'B', 'C'];
+```
+
+</details>
+
+The `fn()` function, for example, ONLY allows you define a callback. This is
+useful for running arbitrary code - if you want to drop down to raw Javascript
+mode, this is how you do it:
+
+```js
+fn(state => {
+  // declare a help function
+  const convertToFhir = item => {
+    /* ... */
+  };
+
+  // Map data into a new format with native Javsacript functions
+  state.transformed = state.data.map(convertToFhir);
+
+  // Always return the state
+  return state;
+});
+```
+
+Many other operations provide a callback argument, in which case, your callback
+will be invoked with state, and most return the final state as a result of the
+operation.
+
+For example, say you fetch some data from a data source and get back a block of
+JSON. Maybe you want to filter this data before passing it to the next
+operation.
+
+You might naively try something like this - but it won't work!
+
+```js
+get('/data'); // writes to state.data
+state.data = state.data.filter(/* ... */); // This is invalid!
+```
+
+You could use another operation, like `fn` or `each` - and often these work
+great!
+
+```js
+get('/data');
+fn(state => {
+  state.data = state.data.filter(/* ... */);
+  return state;
+});
+```
+
+But you can also use a callback function, which is usually a bit neater:
+
+```js
+get('/data', {}, state => {
+  state.data = state.data.filter(/* ... */);
+  return state;
+});
+```
+
+Whatever your callback returns will be used as the input state for the next
+operation (or will be the final state for the job). So remember to ALWAYS return
+state!
+
+Be mindful that some Adaptors will write internal information to state. So you
+should usually `return { ... state }` rather than `return { data: state.data }`.
+
+:::info Remember! Always return state from a callback. :::
+
+## Operations run at the top level
+
+Operations will only work when they are at the top level of your job code, like
+this:
+
+```js
+get('/patients');
+each('$.data.patients', (item, index) => {
+  item.id = `item-${index}`;
+});
+post('/patients', dataValue('patients'));
+```
+
+OpenFn will call your operations in series during workflow execution, ensuring
```

Comment:
@josephjclark should we also link to [this Chaining operations page](https://docs.openfn.org/documentation/build/steps/multiple-operations) and/or edit content on that page?

Reply:
@aleksa-krolls My only problem with the [Chaining operations page](https://docs.openfn.org/documentation/build/steps/multiple-operations) is that it uses `field` and `fields` in that example. And i think that is a pattern we want to deprecate. 

**Two operations approach** 
```
// Prepare a salesforce data model
fn((state) => {
  const { data } = state;

  state.sfMapping = {
    ExternalId: data.form.patient_id,
    Name: data.form.surname,
    "Other Names": data.form.firstName,
    Age__c: data.form.ageInYears,
    Is_Enrolled__c: true,
    Enrollment_Status__c: 3,
  };

  return state;
});

// Upsert patients 
upsert("Patient__c", "ExternalId", (state) => state.sfMapping);
```
**Single operation approach**
```
// Upsert patients
upsert("Patient__c", "ExternalId", ({ data }) => ({
  ExternalId: data.form.patient_id,
  Name: data.form.surname,
  "Other Names": data.form.firstName,
  Age__c: data.form.ageInYears,
  Is_Enrolled__c: true,
  Enrollment_Status__c: 3,
}));
```

I think those examples looks more modern and will cut down explanation on what is `fields`, `filed` and `dataValue`

Reply:
Yeah I'm torn on `fields`. Its kinda neat but also quite an awkward pattern. The single operation approach is way neater - and more familiar to Javascript users. And uses re-usable Javascript skills.

I also find that page to be a bit technical and jargon heavy - I'm really trying to strike a gentler tone here.

But we don't want both pages sitting side by side. Let me take a look at reconciling them.